Added Poison Effect Upon leaving room's area

RealMangorage opened the pull request:
Added the Poison affect so that when players leave their room by somehow getting out of it and into the void of Compact Dimension itll do the following:

Give Them poison and Blindness
Teleports em back to the machine they last went into (if none then it just continues to poison them)

I did it based on what it does in 1.12

Commit:
Merged

Commit:
Test

Commit:
Test

Commit:
Test

Commit:
Test

Commit:
Test

Commit:
Test

Commit:
Test

Commit:
Test

Commit:
Fixed CompactStructureGenerator
 Should now properly remove MachineVoidAir

Commit:
Fixed CompactStructureGenerator
 Should now properly remove MachineVoidAir

Apceniy commented:
But in 1.12 it teleports you out of the machine, not back into it

robotgryphon commented on src/main/java/dev/compactmods/machines/wall/MachineVoidAir.java:
```diff
@@ -0,0 +1,41 @@
+import net.minecraft.world.level.material.Material;
+
+public class MachineVoidAir extends AirBlock {
+    final public static DamageSource DAMAGE_SOURCE = new DamageSource("machinevoidair");
```
Consider underscores/prefix of mod id here

robotgryphon commented on src/main/java/dev/compactmods/machines/wall/MachineVoidAir.java:
```diff
@@ -0,0 +1,41 @@
+            if (player.isCreative()) return;
+
+            if (player.getActiveEffectsMap().containsKey(MobEffects.BLINDNESS))
+                PlayerUtil.teleportPlayerBackToMachine((ServerLevel) pLevel, player);
```
1.12 teleported the player fully out of the machine, prefer that util method

robotgryphon commented on src/main/java/dev/compactmods/machines/upgrade/ChunkloadUpgradeItem.java:
```diff
@@ -10,16 +11,16 @@ public class ChunkloadUpgradeItem extends RoomUpgradeItem {
-
     @Override
     public void fillItemCategory(CreativeModeTab tab, NonNullList<ItemStack> stacks) {
+        if (tab != CompactMachines.COMPACT_MACHINES_ITEMS) return;
```
Unrelated, please revert/open a new PR

robotgryphon commented on src/datagen/java/dev/compactmods/machines/datagen/lang/EnglishLangGenerator.java:
```diff
@@ -120,6 +121,7 @@ protected void addTranslations() {
         add("jei.compactmachines.machines", "Machines are used to make pocket dimensions. Craft a machine and place it in world, then use a Personal Shrinking Device to go inside.");
         add("jei.compactmachines.shrinking_device", "Use the Personal Shrinking Device (PSD) on a machine in order to enter a compact space. " +
                 "You can also right click it in the overworld for more info.");
+        add("death.attack." + MachineVoidAir.DAMAGE_SOURCE.msgId, "%1$s Died from Machine Void Air");
```
Would prefer to align with other death messages, ie "`player` failed to enter the void"

Commit:
Fixed the things that were requested to be fixed.

RealMangorage commented:
Ok I did I fixed em.

Commit:
Added Config for damaging players outside of machine related to MachineVoidAir

robotgryphon commented on src/main/java/dev/compactmods/machines/config/ServerConfig.java:
```diff
@@ -76,6 +78,10 @@ private static void generateConfig() {
                 .comment("Command level required for changing room upgrades.")
                 .defineInRange("upgrades", Commands.LEVEL_GAMEMASTERS, Commands.LEVEL_ALL, Commands.LEVEL_OWNERS);
 
+        ALLOWED_OUTSIDE_MACHINE = builder
+                .comment("If true it'll teleport player out of machine dimension if player is outside of the machine (default: false)")
+                .define("allowoutside", false);
+
```
Needs moved to a new section, this is in the `commands` section. Also, prefer Java casing, `allowOutside`

robotgryphon commented on src/main/java/dev/compactmods/machines/util/PlayerUtil.java:
```diff
@@ -17,11 +18,13 @@
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.GameType;
```
Clean up imports

robotgryphon commented on src/main/java/dev/compactmods/machines/util/PlayerUtil.java:
```diff
@@ -147,6 +150,38 @@ public static void teleportPlayerOutOfMachine(ServerLevel world, @Nonnull Server
         });
     }
 
+    public static void teleportPlayerBackToMachine(ServerLevel world, @Nonnull ServerPlayer serverPlayer) {
+        MinecraftServer serv = world.getServer();
+
+        final LazyOptional<IRoomHistory> history = serverPlayer.getCapability(Capabilities.ROOM_HISTORY);
+
+        if (!history.isPresent()) {
+            howDidYouGetThere(serverPlayer);
+            return;
+        }
+
+        history.ifPresent(hist -> {
+            if (hist.hasHistory()) {
+                final IRoomHistoryItem currentArea = hist.peek();
+                final Level machineLevel = currentArea.getMachine().level(serv);
+                final BlockPos machinePos = currentArea.getMachine().getBlockPosition();
+
+                if(machineLevel.getBlockEntity(machinePos) instanceof CompactMachineBlockEntity tile) {
+                    final var targetRoom = tile.getConnectedRoom();
+                    targetRoom.ifPresent(room -> {
+                        // No need to change history since we just warping player back!
+
+                        try {
+                            teleportPlayerIntoRoom(serv, serverPlayer, room, false);
+                        } catch (MissingDimensionException | NonexistentRoomException e) {
+                            throw new RuntimeException(e);
+                        }
+                    });
+                }
+            }
+        });
+    }
+
```
Evaluating if this is necessary; if the config option isn't an option list between tele back and tele out, this should probably just be removed. If kept, the non-existent history conditional block should fall into returning the player back to their spawn (`teleportPlayerToRespawnOrOverworld`)

robotgryphon commented on src/main/java/dev/compactmods/machines/upgrade/RoomUpgradeItem.java:
```diff
@@ -24,6 +24,8 @@ public RoomUpgradeItem(Properties props) {
     public abstract RoomUpgrade getUpgradeType();
 
+
+
```
Remove rogue spacing please

robotgryphon commented on src/main/java/dev/compactmods/machines/upgrade/ChunkloadUpgradeItem.java:
```diff
@@ -1,5 +1,6 @@
 package dev.compactmods.machines.upgrade;
 
+import dev.compactmods.machines.CompactMachines;
```
Remove unnecessary file changes, to reduce patch size

robotgryphon reviewed (changes requested):
Looking much better; one more cleanup pass and it should be clear

Commit:
Moved ALLOWED_OUTSIDE_MACHINE to correct spot
Cleaned up rogue spacing on RoomUpgradeItem.java
Removed teleportPlayerBackToMachine from PlayerUtil.java
Cleaned up Imports on ChunkloadUpgradeItem.java
